### PR TITLE
feat(workflows): pass ctx

### PR DESCRIPTION
**Pull request by ig-ra**
- pass ctx in session- and event- workflows
- fix passing ctx in resolver
- cleanup/refactoring

### CODE REVIEW

File: internal/backend/dispatcher/eventsworkflow.go [hunk 1]
```diff
@@ -210,12 +210,14 @@ func (d *dispatcher) signalWorkflows(ctx context.Context, event sdktypes.Event)
 	return nil
 }
 
-func (d *dispatcher) eventsWorkflow(ctx workflow.Context, input eventsWorkflowInput) (*eventsWorkflowOutput, error) {
-	logger := workflow.GetLogger(ctx)
-	logger.Info("started events workflow", "event_id", input.EventID)
+func (d *dispatcher) eventsWorkflow(wctx workflow.Context, input eventsWorkflowInput) (*eventsWorkflowOutput, error) {
+	logger := workflow.GetLogger(wctx)
```
Comment (daabr) [inline internal/backend/dispatcher/eventsworkflow.go:214]: Why are you using this, and not just `d.Z`?

**Review by daabr — CHANGES_REQUESTED** [review]
The main thing that this PR is doing is to replace service calls with direct DB access. Unless I'm not aware of a fundamental design change, you should always prefer access via services to direct DB access.

This allows services to manage access security and safety, instead trusting the clients, and implement synchronization logic centrally that clients shouldn't be aware of.

Direct DB access violates design principles like Separation of Concerns, and usually leads to "spaghetti code", and to clashes between clients that require access safeties and other clients that ignore them.

Let's discuss this as a group, to clarify the right approach.

File: internal/backend/dispatcher/resolve.go [hunk 2]
```diff
@@ -32,7 +32,7 @@ func resolveEnv(ctx context.Context, svcs *wf.Services, env string) (sdktypes.En
 			return sdktypes.InvalidEnvID, err
 		}
 
-		p, err := svcs.Projects.GetByName(context.Background(), name)
```
Comment (itayd) [inline internal/backend/dispatcher/resolve.go:35]: I am quite surprised we used Background before.

**Comment by itayd** [comment]
> The main thing that this PR is doing is to replace service calls with direct DB access. Unless I'm not aware of a fundamental design change, you should always prefer access via services to direct DB access.

I don't see it? Maybe it was fixed?

> 
> This allows services to manage access security and safety, instead trusting the clients, and implement synchronization logic centrally that clients shouldn't be aware of.
> 
> Direct DB access violates design principles like Separation of Concerns, and usually leads to "spaghetti code", and to clashes between clients that require access safeties and other clients that ignore them.
> 
> Let's discuss this as a group, to clarify the right approach.

File: internal/backend/dispatcher/resolve.go (hunk 2 above)
Comment (ig-ra) [reply internal/backend/dispatcher/resolve.go:35]: me as well. This PR is fixing this

File: internal/backend/dispatcher/eventsworkflow.go (hunk 1 above)
Comment (ig-ra) [reply internal/backend/dispatcher/eventsworkflow.go:214]: I changed no logic regarding logging except using DB service/interface instead of the services

**Comment by ig-ra** [comment]
> The main thing that this PR is doing is to replace service calls with direct DB access. Unless I'm not aware of a fundamental design change, you should always prefer access via services to direct DB access.
> 
> This allows services to manage access security and safety, instead trusting the clients, and implement synchronization logic centrally that clients shouldn't be aware of.
> 
> Direct DB access violates design principles like Separation of Concerns, and usually leads to "spaghetti code", and to clashes between clients that require access safeties and other clients that ignore them.
> 
> Let's discuss this as a group, to clarify the right approach.

I will remove this specific commit and open another PR for this. Will copy your comment there as well

**Comment by ig-ra** [comment]
> The main thing that this PR is doing is to replace service calls with direct DB access. Unless I'm not aware of a fundamental design change, you should always prefer access via services to direct DB access.
> 
> This allows services to manage access security and safety, instead trusting the clients, and implement synchronization logic centrally that clients shouldn't be aware of.
> 
> Direct DB access violates design principles like Separation of Concerns, and usually leads to "spaghetti code", and to clashes between clients that require access safeties and other clients that ignore them.
> 
> Let's discuss this as a group, to clarify the right approach.

specific commit extracted to https://github.com/autokitteh/autokitteh/pull/511 and removed from this PR

File: internal/backend/sessions/sessionworkflows/workflows.go [hunk 3]
```diff
@@ -208,10 +210,8 @@ func (ws *workflows) sessionWorkflow(wctx workflow.Context, params *sessionWorkf
 	return debug, err
 }
 
-func (ws *workflows) stopped(sessionID sdktypes.SessionID) {
-	z := ws.z.With(zap.String("session_id", sessionID.String()))
-
-	ctx, cancel := withLimitedTimeout(context.Background())
+func (ws *workflows) stopped(ctx context.Context, sessionID sdktypes.SessionID) {
+	ctx, cancel := withLimitedTimeout(ctx)
```
Comment (daabr) [inline internal/backend/sessions/sessionworkflows/workflows.go:214]: Why did you remove the `session_id` tag from the logger?

File: internal/backend/sessions/sessionworkflows/workflows.go [hunk 4]
```diff
@@ -225,20 +225,14 @@ func (ws *workflows) stopped(sessionID sdktypes.SessionID) {
 		}
 	}
 
-	if err := ws.svcs.DB.UpdateSessionState(ctx, sessionID, sdktypes.NewSessionStateStopped(reason)); err != nil {
-		z.Error("update session", zap.Error(err))
-	}
+	_ = ws.updateSessionState(ctx, sessionID, sdktypes.NewSessionStateStopped(reason))
```
Comment (daabr) [inline internal/backend/sessions/sessionworkflows/workflows.go:228]: Why did you remove the error check here?

File: internal/backend/sessions/sessionworkflows/workflows.go [hunk 5]
```diff
@@ -225,20 +225,14 @@ func (ws *workflows) stopped(sessionID sdktypes.SessionID) {
...
 }
 
-func (ws *workflows) errored(sessionID sdktypes.SessionID, err error, prints []string) {
-	z := ws.z.With(zap.String("session_id", sessionID.String()))
-
-	ctx, cancel := withLimitedTimeout(context.Background())
+func (ws *workflows) errored(ctx context.Context, sessionID sdktypes.SessionID, err error, prints []string) {
```
Comment (daabr) [inline internal/backend/sessions/sessionworkflows/workflows.go:231]: Same 2 questions as in `stopped()` above

File: internal/backend/sessions/sessionworkflows/workflows.go [hunk 6]
```diff
@@ -250,6 +244,7 @@ func (ws *workflows) deactivateDrainedDeployment(ctx context.Context, deployment
 			return fmt.Errorf("deployments.get: %w", err)
 		}
 
+		// FIXME: use single query for this and move this to the db layer?
```
Comment (daabr) [inline internal/backend/sessions/sessionworkflows/workflows.go]: Is this meant to be answered and fixed before merging this PR?

**Review by daabr — COMMENTED** [review]
FYI, my comments now are just "comments" and not "request changes" - but I'm guessing the original status won't change until I approve?

File: internal/backend/dispatcher/eventsworkflow.go [hunk 7]
```diff
@@ -210,12 +210,14 @@ func (d *dispatcher) signalWorkflows(ctx context.Context, event sdktypes.Event)
...
 
+	logger.Info("started events workflow", "event_id", input.EventID)
 	z := d.Z.With(zap.String("event_id", input.EventID.String()))
-	event, err := d.Services.Events.Get(context.TODO(), input.EventID)
+
+	ctx := context.Background()
```
Comment (efiShtain) [inline internal/backend/dispatcher/eventsworkflow.go:219]: I think there is a function to get go context from workflow context, maybe it's better then starting a new context ?
could you check please?

File: internal/backend/sessions/sessionworkflows/workflows.go [hunk 8]
```diff
@@ -167,6 +167,8 @@ func (ws *workflows) sessionWorkflow(wctx workflow.Context, params *sessionWorkf
 		ScheduleToCloseTimeout: ws.cfg.Temporal.LocalScheduleToCloseTimeout,
 	})
 
+	ctx := context.Background()
```
Comment (efiShtain) [inline internal/backend/sessions/sessionworkflows/workflows.go:170]: why new context ?

File: internal/backend/sessions/sessionworkflows/workflows.go (hunk 4 above)
Comment (ig-ra) [reply internal/backend/sessions/sessionworkflows/workflows.go:228]: I don't think I did. 
1. There was duplicated code and already existing method which updatesSessionState and logs
2. On error the only thing happened is log, and it's done now in single method - updateSessionState

File: internal/backend/sessions/sessionworkflows/workflows.go (hunk 5 above)
Comment (ig-ra) [reply internal/backend/sessions/sessionworkflows/workflows.go:231]: answered above

File: internal/backend/dispatcher/eventsworkflow.go (hunk 1 above)
Comment (ig-ra) [reply internal/backend/dispatcher/eventsworkflow.go:214]: resolve?

File: internal/backend/sessions/sessionworkflows/workflows.go (hunk 6 above)
Comment (ig-ra) [reply internal/backend/sessions/sessionworkflows/workflows.go]: no, since this PR isn't handling this. I only added FIXME since  I think it should be done differently. Should I change it to TODO?

File: internal/backend/sessions/sessionworkflows/workflows.go (hunk 8 above)
Comment (ig-ra) [reply internal/backend/sessions/sessionworkflows/workflows.go:170]: explain please?

File: internal/backend/dispatcher/eventsworkflow.go (hunk 7 above)
Comment (ig-ra) [reply internal/backend/dispatcher/eventsworkflow.go:219]: maybe. But in old code new context was started many times when it needed to be passed. I'm just going 1 step a time - I create it once and pass it through functions. Maybe we could do it event better and obtain it from the workflow context. At least this change was checked by me when using tagging components, and using context from workflow - isn't. So I prefer to stay meanwhile with that

File: internal/backend/sessions/sessionworkflows/workflows.go (hunk 8 above)
Comment (efiShtain) [reply internal/backend/sessions/sessionworkflows/workflows.go:170]: my bad, it was collapsed and I didn't see it's a new function

resolving

**Review by daabr — APPROVED** [review]
